src/main/sensors/rpm_filter.c

```diff
@@ -0,0 +1,204 @@
+}
+
+static void rpmNotchFilterInit(rpmNotchFilter_t* filter, int harmonics, int minHz, int q, float looptime)
+{
+    filter->harmonics = harmonics;
+    filter->minHz = minHz;
+    filter->q = q / 10.0f;
+    filter->loopTime = looptime;
+
+    for (int axis = 0; axis < XYZ_AXIS_COUNT; axis++) {
+        for (int motor = 0; motor < getMotorCount(); motor++) {
+            for (int i = 0; i < harmonics; i++) {
+                biquadFilterInit(
+                    &filter->notch[axis][motor][i], minHz * i, looptime, filter->q, FILTER_NOTCH);
+            }
+        }
+    }
+}
+
+void rpmFilterInit(const rpmFilterConfig_t *config)
+{
+    pidLooptime = gyro.targetLooptime * pidConfig()->pid_process_denom;
+    numberRpmNotchFilters = 0;
+    if (config->gyro_rpm_notch_harmonics) {
+        gyroFilter = &filters[numberRpmNotchFilters++];
+        rpmNotchFilterInit(gyroFilter, config->gyro_rpm_notch_harmonics,
+                           config->gyro_rpm_notch_min, config->gyro_rpm_notch_q, gyro.targetLooptime);
+    }
+    if (config->dterm_rpm_notch_harmonics) {
+        dtermFilter = &filters[numberRpmNotchFilters++];
+        rpmNotchFilterInit(dtermFilter, config->dterm_rpm_notch_harmonics,
+                           config->dterm_rpm_notch_min, config->dterm_rpm_notch_q, pidLooptime);
+    }
+
+    for (int i = 0; i < getMotorCount(); i++) {
+        pt1FilterInit(&rpmFilters[i], pt1FilterGain(RPM_MOTOR_FILTER_CUTOFF, gyro.targetLooptime));
+    }
+
+    const float secondsPerMinute = 60.0f;
+    const float erpmPerLsb = 100.0f;
+    erpmToHz = erpmPerLsb / secondsPerMinute  / (motorConfig()->motorPoleCount / 2.0f);
+
+    const float minUpdateT = 0.001f;
+    const float loopIterationsPerUpdate = minUpdateT / (pidLooptime * 1e-6f);
+    numberFilters = getMotorCount() * (filters[0].harmonics + filters[1].harmonics);
+    const float filtersPerLoopIteration = numberFilters / loopIterationsPerUpdate;
+    filterUpdatesPerIteration = rintf(filtersPerLoopIteration + 0.49f);
+}
+
+static float applyFilter(rpmNotchFilter_t* filter, int axis, float value)
+{
+    if (filter == NULL) {
+        return value;
+    }
+    for (int motor = 0; motor < getMotorCount(); motor++) {
+        for (int i = 0; i < filter->harmonics; i++) {
+            value = biquadFilterApplyDF1(&filter->notch[axis][motor][i], value);
+        }
+    }
+    return value;
+}
+
+float rpmFilterGyro(int axis, float value)
+{
+    return applyFilter(gyroFilter, axis, value);
+}
+
+float rpmFilterDterm(int axis, float value)
+{
+    return applyFilter(dtermFilter, axis, value);
+}
+
+static float motorFrequency[MAX_SUPPORTED_MOTORS];
+
+void rpmFilterUpdate()
+{
+    if (gyroFilter == NULL && dtermFilter == NULL) {
+        return;
+    }
+
+    for (int motor = 0; motor < getMotorCount(); motor++) {
+        filteredMotorErpm[motor] = pt1FilterApply(&rpmFilters[motor], getDshotTelemetry(motor));
+        DEBUG_SET(DEBUG_RPM_FILTER, motor, motorFrequency[motor]);
```

Comment:
This will result in an overrun of the `debug` array for configurations with more than 4 motors.

In general, since this will scale in the number of motors that have to be processed, it might make sense to disable this in `validateAndFixConfig()` if more than 4 motors are configured otherwise those users with sedecacopters might experience high task load. :wink:

Reply:
Good catch on the debug. I'm not going to address the sedecacopters just yet. But it might make sense in the future to make the minimum update time configurable so it can run slower on slower boards.